common/kvstore/tablestore/table_store_wrapper.go

```diff
@@ -0,0 +1,353 @@
+		if onDiskSchema != currentSchemaVersion {
+			// In the future if we change schema versions, we may need to write migration code here.
+			return nil, fmt.Errorf(
+				"incompatible schema version: code is at version %d, data on disk is at version %d",
+				currentSchemaVersion, onDiskSchema)
+		}
+
+		highestTableID, err = loadNamespaceTable(base, namespaceTable, tableIDMap, tableMap, tableIdSet)
+		if err != nil {
+			return nil, fmt.Errorf("error loading namespace table: %w", err)
+		}
+	}
+
+	store := &tableStore{
+		logger:         logger,
+		base:           base,
+		tableIDMap:     tableIDMap,
+		tableMap:       tableMap,
+		metadataTable:  metadataTable,
+		namespaceTable: namespaceTable,
+		highestTableID: highestTableID,
+	}
+
+	err = store.handleIncompleteDeletion()
+	if err != nil {
+		return nil, fmt.Errorf("error handling incomplete deletion: %w", err)
+	}
+
+	return store, nil
+}
+
+// loadNamespaceTable loads the namespace table from disk into the given map. Returns the highest table ID found.
+func loadNamespaceTable(
+	base kvstore.Store,
+	namespaceTable kvstore.Table,
+	tableIDMap map[string]uint32,
+	tableMap map[uint32]kvstore.Table,
+	tableIdSet map[uint32]bool) (int64, error) {
+
+	highestTableID := int64(-1)
+
+	it, err := namespaceTable.NewIterator(nil)
+	if err != nil {
+		return -1, fmt.Errorf("error creating namespace table iterator: %w", err)
+	}
+	defer it.Release()
+
+	for it.Next() {
+		keyBytes := it.Key()
+		valueBytes := it.Value()
+
+		tableID := binary.BigEndian.Uint32(keyBytes)
+		tableName := string(valueBytes)
+		tableIdSet[tableID] = true
+		tableIDMap[tableName] = tableID
+		tableMap[tableID] = newTableView(base, tableName, tableID)
+
+		if int64(tableID) > highestTableID {
+			highestTableID = int64(tableID)
+		}
+	}
+
+	return highestTableID, nil
+}
+
+// This method handles cleanup of incomplete deletions. Since deletion of a table requires multiple operations that
+// are not atomic in aggregate, it is possible that a table deletion may have been started without being completed.
+// This method makes sure that any such incomplete deletions are completed.
+func (t *tableStore) handleIncompleteDeletion() error {
+	deletionTableNameBytes, err := t.metadataTable.Get([]byte(metadataDeletionKey))
+	if errors.Is(err, kvstore.ErrNotFound) {
+		// No deletion in progress, nothing to do.
+		return nil
+	}
+
+	deletionTableName := string(deletionTableNameBytes)
+	t.logger.Errorf("found incomplete deletion of table %s, completing deletion", deletionTableName)
+
+	return t.DropTable(deletionTableName)
+}
+
+// Get the prefix for the given table ID and prefix length.
+func getPrefix(tableID uint32) []byte {
+	bytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(bytes, tableID)
+	return bytes
+}
+
+// GetTable gets the table with the given name. If the table does not exist, it is first created.
+func (t *tableStore) GetTable(name string) (kvstore.Table, error) {
```

Comment:
I don't see this TableStore is thread-safe (which is required per interface)

Reply:
In the interface, this method is specifically documented not to be thread safe (although the godoc in the implementation doesn't copy the warning). The code to make this thread simple to write, but I'm concerned about adding locks that could slow other operations. Do you think it's worth adding locks so that these methods are safe to call concurrently?

```
	// GetTable gets the table with the given name. If the table does not exist, it is first created.
	//
	// WARNING: this method is not thread safe with respect to any other methods in this interface or
	// any methods on any Table objects associated with this store.
	GetTable(name string) (Table, error)
```

Reply:
I think it needs to, otherwise other methods like GetTableCount (which is supposed to be thread safe) won't be safe as they share the states.

Reply:
There are now no unsafe methods (due to schema pattern)